codelabs/custom_generator/custom_generator.md

```diff
@@ -0,0 +1,379 @@
+```
+
+Note: you must include your custom code *after* including the Blockly library.
+
+### Blocks
+
+For this codelab you will use two custom blocks, as well as five blocks from Blockly's standard set.
+
+The custom blocks represent the *Object* and *Member* sections of the JSON specification.
+
+The blocks are:
+- object
+- member
+- math_number
+- text
+- logic_boolean
+- logic_null
+- lists_create_with
+
+### Custom block definitions
+
+Copy this code into `custom_generator.js` to define the two custom blocks.
+
+```js
+Blockly.defineBlocksWithJsonArray([{
+  "type": "object",
+  "message0": "{ %1 %2 }",
+  "args0": [
+    {
+      "type": "input_dummy"
+    },
+    {
+      "type": "input_statement",
+      "name": "MEMBERS"
+    }
+  ],
+  "output": null,
+  "colour": 230,
+},
+{
+  "type": "member",
+  "message0": "%1 %2 %3",
+  "args0": [
+    {
+      "type": "field_input",
+      "name": "MEMBER_NAME",
+      "text": ""
+    },
+    {
+      "type": "field_label",
+      "name": "COLON",
+      "text": ":"
+    },
+    {
+      "type": "input_value",
+      "name": "MEMBER_VALUE"
+    }
+  ],
+  "previousStatement": null,
+  "nextStatement": null,
+  "colour": 230,
+}]);
+```
+
+### Toolbox definition
+
+Next, define your toolbox in XML. For this example we have a flyout-only toolbox with seven blocks in it.
+
+Copy this code into `custom_generator.js`:
+
+```js
+var codelabToolbox = `
+<xml id="toolbox" style="display: none">
+<block type="object"/>
+<block type="member"></block>
+<block type="math_number"><field name="NUM">0</field></block>
+<block type="text"><field name="TEXT"/></block>
+<block type="logic_boolean"><field name="BOOL">TRUE</field></block>
+<block type="logic_null"/>
+<block type="lists_create_with"><mutation items="3"/></block></xml>
+`
+```
+
+Then update the options struct in the playgound to use your new toolbox:
+
+```js
+var defaultOptions = {
+  comments: true,
+  // ...
+  toolbox: codelabToolbox,
+  // ...
+};
+```
+
+## The basics
+
+The first step is to define and call your generator.
+
+### Create your generator
+
+A custom generator is simply an instance of `Blockly.Generator`. Call the constructor, passing in your generator's name, and store the result.
+
+```js
+const myGenerator = new Blockly.Generator('Codelab');
+```
+
+### Generate code
+
+Add a button to the playground with the following callback:
+
+```js
+const generatedCode = myGenerator.workspaceToCode(Blockly.getMainWorkspace());
+console.log(generatedCode);
+```
+
+Note: You must use `workspaceToCode` as the entry point for code generation, passing in the workspace. `workspaceToCode` calls `init` and `finish` to do any additional setup and teardown that your generator needs.
+
+Other functions such as `blockToCode` are public so that you can use them in your per-block generator definitions, but must not be used as the entry point.
+
+### Test it
+
+Put a number block on the workspace, then click your button to call your generator.
```

Comment:
A screenshot here to help follow along

Reply:
Also, what button, and what's it calling? Perhaps refer to the button in the previous step as the "Generate" button, and then something like: "then click the Generate button to call your generator's workspaceToCode method"

Reply:
Screenshot: definitely. I'm adding a todo for now, because it'll depend on which playground I'm using.

Reply:
I ended up with no screenshot here because it didn't feel helpful, but instead I added much more explicit instructions on how and where to add the button and its callback.